### PR TITLE
feat(preferences): add app settings for new proxy options COMPASS-8141

### PR DESCRIPTION
**Pull request by addaleax**
This isn't hooked up to any actual proxy logic yet, to be clear 🙂

### CODE REVIEW

File: packages/compass-preferences-model/src/preferences-persistent-storage.spec.ts [hunk 1]
```diff
@@ -22,7 +28,7 @@ describe('PersistentStorage', function () {
   });
 
   afterEach(async function () {
-    await fs.rmdir(tmpDir, { recursive: true });
+    await fs.rm(tmpDir, { recursive: true, force: true });
```
Comment (addaleax) [inline packages/compass-preferences-model/src/preferences-persistent-storage.spec.ts:31]: Drive-by: This fixes a deprecation warning

File: packages/compass-preferences-model/src/preferences-schema.ts [hunk 2]
```diff
@@ -760,6 +780,52 @@ export const storedUserPreferencesProps: Required<{
...
+      merge({ remainder, secrets }) {
+        return proxyOptionsToProxyPreference(
+          mergeProxySecrets({
+            ...JSON.parse(remainder),
+            secrets: JSON.parse(secrets),
+          })
+        );
+      },
+    },
+  },
```
Comment (addaleax) [inline packages/compass-preferences-model/src/preferences-schema.ts:827]: This is one of the main two things that I'd like feedback on – how do we feel about this approach to separating secrets from non-secrets in preferences? This is unfortunately not something I considered a lot in the tech design, and I did try another approach first (splitting these into separate preference entries and doing the encryption/decryption in the code that accesses these), but so far it's something I think works pretty well

File: packages/compass-settings/src/stores/settings.ts [hunk 3]
```diff
@@ -173,21 +229,37 @@ export const changeFieldValue = <K extends keyof UserConfigurablePreferences>(
   };
 };
 
-export const openModal = (): SettingsThunkAction<Promise<void>> => {
+export const openModal = (
+  tab?: SettingsTabId
+): SettingsThunkAction<Promise<void>, OpenSettingsModalAction> => {
   return async (dispatch) => {
-    dispatch({ type: ActionTypes.OpenSettingsModal });
+    dispatch({ type: ActionTypes.OpenSettingsModal, tab });
```
Comment (addaleax) [inline packages/compass-settings/src/stores/settings.ts:236]: ... and this the other thing, I guess, although I don't know much about what potential alternatives would look like – this is moving the "open tab" state to redux from the React component, and this will enable us to open a specific tab in the modal from other parts of Compass (which is planned for the next PR)

File: packages/compass-e2e-tests/helpers/compass.ts [hunk 4]
```diff
@@ -472,7 +472,8 @@ async function getCompassExecutionParameters(): Promise<{
   );
   const binary = testPackagedApp
     ? getCompassBinPath(await getCompassBuildMetadata())
-    : require('electron');
+    : // eslint-disable-next-line @typescript-eslint/no-var-requires
+      (require('electron') as unknown as string);
```
Comment (addaleax) [inline packages/compass-e2e-tests/helpers/compass.ts:476]: Really unsure why this started popping up as an issue for me tbh – feels like it should always have been complaining about this

File: packages/compass-e2e-tests/tests/auto-connect.test.ts [hunk 5]
```diff
@@ -245,7 +245,7 @@ describe('Automatically connecting from the command line', function () {
       await browser.waitForConnectionResult('success');
       await browser.execute(() => {
         // eslint-disable-next-line @typescript-eslint/no-var-requires
-        require('electron').ipcRenderer.call('test:show-connect-window');
+        require('electron').ipcRenderer.invoke('test:show-connect-window');
```
Comment (addaleax) [inline packages/compass-e2e-tests/tests/auto-connect.test.ts:248]: Similarly here, this is unrelated but was required for me to be able to run e2e tests at all

File: packages/compass-settings/src/stores/settings.ts [hunk 6]
```diff
@@ -52,9 +63,43 @@ type SaveSettingsAction = {
...
+  | SettingsFetchedAction
+  | SaveSettingsAction
+  | SettingsFetchStartAction
+  | OpenSettingsModalAction
+  | SelectTabAction
+  | CloseSettingsModalAction
+  | ChangeFieldValueAction<keyof UserConfigurablePreferences>;
 
-export const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
+export const reducer: Reducer<State, Actions> = (
```
Comment (gribnoysup) [inline packages/compass-settings/src/stores/settings.ts:99]: I know it's a pain in the butt, but Redux docs [recommend against unions](https://redux.js.org/usage/usage-with-typescript#avoid-action-type-unions) for action type and we generally follow this rule by using [isAction](https://github.com/mongodb-js/compass/blob/9b0315879799f8d884944b6779b44976609f2e2a/packages/atlas-service/src/store/atlas-signin-reducer.ts#L8-L13) helper

File: packages/compass-settings/src/stores/settings.ts [hunk 7]
```diff
@@ -67,13 +112,21 @@ export const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
...
+        tab: action.tab ?? state.tab,
       };
     case ActionTypes.CloseSettingsModal:
       return {
         ...state,
         isModalOpen: false,
+        tab: undefined,
       };
     case ActionTypes.ChangeFieldValue:
+      if (state.loadingState !== 'ready') return { ...state };
```
Comment (gribnoysup) [inline packages/compass-settings/src/stores/settings.ts:129]: Nit: no need for shallow cloning if state is not changed

```suggestion
      if (state.loadingState !== 'ready') return state;
```

File: packages/compass-settings/src/stores/settings.ts (hunk 3 above)
Comment (gribnoysup) [reply packages/compass-settings/src/stores/settings.ts:236]: Yeah, that's definitely the way to go about it. We can also now change the welcome modal to link directly to privacy tab if we want to

File: packages/compass-settings/src/stores/settings.ts [hunk 8]
```diff
@@ -173,21 +229,37 @@ export const changeFieldValue = <K extends keyof UserConfigurablePreferences>(
...
 };
 
-export const closeModal = (): SettingsThunkAction<void> => {
+export const selectTab = (
+  tab?: SettingsTabId
+): SettingsThunkAction<void, SelectTabAction> => {
+  return (dispatch) => {
+    dispatch({ type: ActionTypes.SelectTab, tab });
+  };
+};
```
Comment (gribnoysup) [inline packages/compass-settings/src/stores/settings.ts:247]: Small nit: doesn't need to be a thunk action. Nothing wrong with it, just you usually might want to use those only if you need to dispatch multiple events from one action, access current state before dispatching, or have access to the thunk "extra arg". None of these are the case here

```suggestion
): SelectTabAction => {
  return { type: ActionTypes.SelectTab, tab }
};
```

File: packages/compass-settings/src/stores/settings.ts (hunk 6 above)
Comment (addaleax) [reply packages/compass-settings/src/stores/settings.ts:99]: Ah, right. I literally forgot that that's what we do in Compass tbh 🙈 Just saw that `action` wasn't typed and wanted to fix that, but I'll gladly remove this again and use `isAction` instead, will do that in the next commit.

It does make me wonder, should we maybe prefer typing our reducers with `Reducer<State, Action>` (instead of the implied default `Reducer<State, AnyAction>`)? That would still make sure that we're not accidentally relying on the `any`-ness of the `action` parameter we're using

File: packages/compass-settings/src/stores/settings.ts (hunk 8 above)
Comment (addaleax) [reply packages/compass-settings/src/stores/settings.ts]: Sure – done in b3a9e6f02dacd0e33a2a6b8eb959bf1aca436534 👍

File: packages/compass-settings/src/stores/settings.ts (hunk 3 above)
Comment (addaleax) [reply packages/compass-settings/src/stores/settings.ts:249]: Oh, yeah, that makes a ton of sense 🙂 Done in 20a2084c037a1c40e2e6b4ab9f92c01c814f222d!

File: packages/compass-settings/src/stores/settings.ts (hunk 7 above)
Comment (addaleax) [reply packages/compass-settings/src/stores/settings.ts]: Done in 8236817e9e55c318f7e578a80722b8310c0f7742 :+1:

File: packages/compass-settings/src/stores/settings.ts (hunk 6 above)
Comment (addaleax) [reply packages/compass-settings/src/stores/settings.ts]: Done in 8236817e9e55c318f7e578a80722b8310c0f7742 :+1:

File: packages/compass-settings/src/stores/settings.ts [hunk 9]
```diff
@@ -67,13 +112,21 @@ export const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
...
+    case ActionTypes.SelectTab:
+      return {
+        ...state,
+        tab: action.tab ?? state.tab,
       };
     case ActionTypes.CloseSettingsModal:
       return {
         ...state,
         isModalOpen: false,
+        tab: undefined,
```
[the remark this answers is not in the document]
Comment (addaleax) [reply packages/compass-settings/src/stores/settings.ts]: Thanks for explaining! Absolutely makes sense, done in 2d770513c667bc5c4a36ea6a214a95d442951fbd :)

File: packages/compass-preferences-model/src/preferences-schema.ts [hunk 10]
```diff
@@ -5,6 +5,14 @@ import {
   featureFlags,
 } from './feature-flags';
 import { parseRecord } from './parse-record';
+import {
+  extractProxySecrets,
+  mergeProxySecrets,
+} from '@mongodb-js/devtools-proxy-support';
```
Comment (addaleax) [inline packages/compass-preferences-model/src/preferences-schema.ts:11]:
```suggestion
} from '@mongodb-js/devtools-proxy-support/proxy-options';
```

added a commit to https://github.com/mongodb-js/devtools-shared/pull/451 to make these parts browser-compatible

File: configs/tsconfig-compass/tsconfig.common.json [hunk 11]
```diff
@@ -1,3 +1,6 @@
 {
-  "extends": "@mongodb-js/tsconfig-devtools/tsconfig.common.json"
+  "extends": "@mongodb-js/tsconfig-devtools/tsconfig.common.json",
+  "compilerOptions": {
+    "moduleResolution": "node16"
+  }
```
Comment (addaleax) [inline configs/tsconfig-compass/tsconfig.common.json:5]: ... so that the `'@mongodb-js/devtools-proxy-support/proxy-options'` import works (which in turn also causes the small change to the `mongodb-explain-compat` package.json).

Comment (gribnoysup) [reply configs/tsconfig-compass/tsconfig.common.json]: Woah, last time I tried to turn it on, a bunch of other things broke, which are maybe gone now? This means that we can also remove weird workarounds for provider re-exports, awesome!

Comment (addaleax) [reply configs/tsconfig-compass/tsconfig.common.json]: Fwiw, `npm run bootstrap` passed just fine – but this failed during the e2e test phase. Probably not hard to fix if we want to, though

**Review by gribnoysup — APPROVED** [review]
